gemini/sample-apps/genai-mlops-tune-and-eval/local/pipeline.py

```diff
@@ -0,0 +1,253 @@
+            MetricPromptTemplateExamples.Pointwise.GROUNDEDNESS,
+        ],
+        experiment_name: str = experiment_name,
+    ) -> object:
+        """
+        Takes the instruction, context and a variable number of corresponding generated responses, and returns the pointwise evaluation metrics
+        for each of the provided metrics. For this example the metrics are Q & A related, however the full list can be found on the website:
+        https://cloud.google.com/vertex-ai/generative-ai/docs/models/online-pipeline-services
+        """
+
+        instructions = [instruction] * len(responses)
+
+        contexts = [context] * len(responses)
+
+        eval_dataset = pd.DataFrame(
+            {
+                "instruction": instructions,
+                "context": contexts,
+                "response": responses,
+            }
+        )
+
+        eval_task = EvalTask(
+            dataset=eval_dataset, metrics=eval_metrics, experiment=experiment_name
+        )
+        results = eval_task.evaluate(
+            prompt_template="{instruction} \n {context}",
+            experiment_run_name="gemini-qa-pointwise-" + str(uuid.uuid4()),
+        )
+        (results.metrics_table.columns)
+        return results
+
+    def rank_responses(instruction: str, context: str, responses: list[str]) -> tuple:
+        """
+        Takes the instruction, context and a variable number of responses as input, and returns the best performing response as well as its associated
+        human readable pointwise quality metrics for the configured criteria in the above functions.
+        The process consists of two steps:
+        1. Selecting the best response by using Pairwise comparisons between the responses for the user specified metric ( e.g. Q & A)
+        2. Doing pointwise evaluation of the best response and returning human readable quality metrics and explanation along with the best response.
+        """
+        cmp_f = partial(
+            pairwise_greater, instruction, context, project, location, experiment_name
+        )
+        cmp_greater = partial(greater, cmp_f)
+
+        pairwise_best_response = max(responses, key=functools.cmp_to_key(cmp_greater))
+        pointwise_metric = pointwise_eval(
+            instruction, context, [pairwise_best_response]
+        )
+        qa_metrics = pointwise_metric.metrics_table[
+            [
+                col
+                for col in pointwise_metric.metrics_table.columns
+                if ("question_answering" in col) or ("groundedness" in col)
+            ]
+        ].to_dict("records")[0]
+
+        return pairwise_best_response, qa_metrics
+
+    # Compare response from baseline model to candidate model to see which is better
+    baseline_model = GenerativeModel(
+        baseline_model_endpoint,
+        generation_config={
+            "temperature": 0.4,
+            "max_output_tokens": 512,
+        },
```

Comment:
Can you use the `GenerationConfig` class?

Reply:
Done